### PR TITLE
Modeling updates

### PR DESCRIPTION
**Pull request by jekbradbury**
A handful of small changes, plus addition of dropout and layernorm. Supersedes/stacked on https://github.com/rxwei/DeepLearning/pull/3.

### CODE REVIEW

File: Sources/DeepLearning/Layer.swift [hunk 1]
```diff
@@ -268,3 +270,94 @@ public struct AvgPool2D<Scalar>: Layer
...
+        } else {
+            return applyInference(to: input)
+        }
+    }
+
+    public func _vjpApplied(to input: Tensor<Scalar>) ->
+        (Tensor<Scalar>, (Tensor<Scalar>) ->
+            (Dropout<Scalar>.CotangentVector, Tensor<Scalar>)) {
+        if mode.training {
+            return Swift.valueWithPullback(at: self, input) {
```
Comment (dan-zheng) [inline Sources/DeepLearning/Layer.swift]: Wonder if the [method-style differential operator](https://github.com/apple/swift/blob/a6a2790d64f8512e93e4ca75b0c1f2e0d70d48f3/stdlib/public/core/AutoDiff.swift#L215) works?
```swift
        if mode.training {
            return valueWithPullback(at: input) {
                $0.applyTraining(to: $1)
            }
        else {
            return valueWithPullback(at: input) {
                $0.applyInference(to: $1)
            }
        }
```

File: Sources/DeepLearning/Layer.swift [hunk 2]
```diff
@@ -268,3 +270,94 @@ public struct AvgPool2D<Scalar>: Layer
...
+    }
+}
+
+@_fixed_layout
+public struct Dropout<Scalar>: Layer
+    where Scalar: BinaryFloatingPoint & Differentiable & TensorFlowScalar,
+          Scalar.RawSignificand : FixedWidthInteger {
+    @noDerivative public let dropProb: Double
+    @noDerivative public let mode: ModeRef
+    var _unused: Tensor<Scalar>
```
Comment (dan-zheng) [inline Sources/DeepLearning/Layer.swift:325]: Is `_unused` a workaround for [TF-8](https://bugs.swift.org/browse/TF-8)?
If so, could you please add a comment? (`// Blocked by TF-8.`)

**Review by dan-zheng — APPROVED** [review]

File: Sources/DeepLearning/Layer.swift [hunk 3]
```diff
@@ -268,3 +270,94 @@ public struct AvgPool2D<Scalar>: Layer
...
+
+    public init(dropProbability: Double, modeRef: ModeRef) {
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Just "probability" would be better.

```suggestion
    public init(probability: Double, modeRef: ModeRef) {
```

File: Sources/DeepLearning/Layer.swift [hunk 4]
```diff
@@ -268,3 +270,94 @@ public struct AvgPool2D<Scalar>: Layer
...
+        self.dropProb = dropProbability
+        self.mode = modeRef
+        self._unused = Tensor<Scalar>(Scalar(0))
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: `BinaryFloatingPoint` is convertible from integer literal.

```suggestion
        self._unused = Tensor<Scalar>(0)
```

File: Sources/DeepLearning/Layer.swift [hunk 5]
```diff
@@ -268,3 +270,94 @@ public struct AvgPool2D<Scalar>: Layer
...
+    }
+
+    @differentiable(wrt: (self, input))
+    public func applyTraining(to input: Tensor<Scalar>) -> Tensor<Scalar> {
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Why make this and `applyInference(to:)` public? If `applied(to:)` with mode toggling worked, these should be private.

File: Sources/DeepLearning/Layer.swift [hunk 6]
```diff
@@ -268,3 +270,94 @@ public struct AvgPool2D<Scalar>: Layer
...
+        let keepProbability = Scalar(1.0 - dropProbability)
+        return self * Tensor(keepMask) / Tensor(keepProbability)
+    }
+}
+
+@_fixed_layout
+public struct Dropout<Scalar>: Layer
+    where Scalar: BinaryFloatingPoint & Differentiable & TensorFlowScalar,
+          Scalar.RawSignificand : FixedWidthInteger {
+    @noDerivative public let dropProb: Double
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]:
```suggestion
    @noDerivative public let probability: Double
```

File: Sources/DeepLearning/Layer.swift [hunk 7]
```diff
@@ -52,12 +52,12 @@ public extension Layer {
 /// between training and inference shares an instance of ModeRef so it doesn't
 /// need to be toggled or threaded through in more than one place.
 public class ModeRef {
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Avoid using short names unless it's a word of art. In this case, this should be renamed to something like `InferenceModeIndicator`.

Comment (jekbradbury) [reply Sources/DeepLearning/Layer.swift]: I think I’m going to adopt the Keras terminology (“learning phase”), although I also like “mode” as a noun

Comment (rxwei) [reply Sources/DeepLearning/Layer.swift]: +1 to "learning phase", but I would add a word like "indicator" to indicate its reference semantics.

File: Sources/DeepLearning/Layer.swift (hunk 3 above)
Comment (jekbradbury) [reply Sources/DeepLearning/Layer.swift]: Unfortunately, libraries haven’t standardized completely on whether the relevant scalar represents the “drop probability” or the “keep probability” so I’m not sure “probability” alone is necessarily enough

Comment (rxwei) [reply Sources/DeepLearning/Layer.swift]: My sense is that libraries that do mean "keep probability" have the word "keep" in the variable, e.g. `keep_prob`. Those meaning "drop probability" use "rate", "dropout rate" (which repeats the type name), or just "p". So I think the word "probability" in this type clearly indicates "dropout probability" without ambiguity.

File: Sources/DeepLearning/Layer.swift (hunk 1 above)
Comment (jekbradbury) [reply Sources/DeepLearning/Layer.swift]: It didn't when I first tried it, because of a mismatch between curried and noncurried signatures. Can try again though!

Comment (jekbradbury) [reply Sources/DeepLearning/Layer.swift]: Seems to work now. I was probably doing something wrong before.

File: Sources/DeepLearning/Layer.swift [hunk 8]
```diff
@@ -268,3 +270,99 @@ public struct AvgPool2D<Scalar>: Layer
...
+        let mean = input.mean(alongAxes: axis)
+        let variance = input.variance(alongAxes: axis)
+        let inv = rsqrt(variance + epsilon) * scale
+        return (input - mean) * inv + offset
+    }
+}
+
+public extension Tensor where Scalar : BinaryFloatingPoint,
+                       Scalar.RawSignificand : FixedWidthInteger {
+    @differentiable(wrt: (self) where Scalar : Differentiable)
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]:
```suggestion
    @differentiable(wrt: self where Scalar : Differentiable)
```

File: Sources/DeepLearning/Layer.swift [hunk 9]
```diff
@@ -268,3 +270,99 @@ public struct AvgPool2D<Scalar>: Layer
...
+    public func applied(to input: Tensor<Scalar>) -> Tensor<Scalar> {
+        let mean = input.mean(alongAxes: axis)
+        let variance = input.variance(alongAxes: axis)
+        let inv = rsqrt(variance + epsilon) * scale
+        return (input - mean) * inv + offset
+    }
+}
+
+public extension Tensor where Scalar : BinaryFloatingPoint,
+                       Scalar.RawSignificand : FixedWidthInteger {
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Align with the last where clause.

File: Sources/DeepLearning/Layer.swift [hunk 10]
```diff
@@ -268,3 +270,99 @@ public struct AvgPool2D<Scalar>: Layer
...
+    @differentiable(wrt: (self) where Scalar : Differentiable)
+    func droppedOut(withProbability probability: Double) -> Tensor {
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: I'd suggest `droppingOut(probability:)`.
- Using the past participle to mean dropping some elements out sounds a bit odd.
- `withProbability` is a slightly too long for a label. In these cases we'd often choose to drop a needless word.

File: Sources/DeepLearning/Layer.swift [hunk 11]
```diff
@@ -268,3 +270,99 @@ public struct AvgPool2D<Scalar>: Layer
...
+        learningPhaseIndicator: LearningPhaseIndicator
+    ) {
+        self.probability = probability
+        self.learningPhaseIndicator = learningPhaseIndicator
+        // Workaround for TF-8
+        self._unused = Tensor<Scalar>(0)
+    }
+
+    @differentiable(wrt: (self, input))
+    func applyTraining(to input: Tensor<Scalar>) -> Tensor<Scalar> {
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Would `private` work?

**Review by rxwei — APPROVED** [review]

File: Sources/DeepLearning/Layer.swift [hunk 12]
```diff
@@ -51,13 +51,14 @@ public extension Layer {
 /// In typical uses, every layer in a model that has behavior which differs
 /// between training and inference shares an instance of ModeRef so it doesn't
 /// need to be toggled or threaded through in more than one place.
-public class ModeRef {
-    var training: Bool = true
+public class LearningPhaseIndicator {
+    public var training: Bool = true
+    public init() { }
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]:
```suggestion
    public init() {}
```

File: Sources/DeepLearning/Layer.swift [hunk 13]
```diff
@@ -51,13 +51,14 @@ public extension Layer {
...
 }
 
 /// A mutable, shareable reference to a tensor
 public class Parameter<T : TensorFlowScalar> {
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Nit: In non-stdlib code, not add a space before a colon in all cases (including the inheritance syntax).
```suggestion
public class Parameter<T: TensorFlowScalar> {
```

**Review by rxwei — APPROVED** [review]
